core/iwasm/aot/aot_runtime.c

```diff
@@ -1444,8 +1465,26 @@ aot_call_function(WASMExecEnv *exec_env, AOTFunctionInstance *function,
         }
 #endif
 
-        ret = invoke_native_internal(exec_env, function->u.func.func_ptr,
-                                     func_type, NULL, NULL, argv, argc, argv);
+#ifdef OS_ENABLE_BLOCK_INSN_INTERRUPT
+        wasm_exec_env_push_jmpbuf(exec_env, &jmpbuf_node);
+        wasm_runtime_set_exec_env_tls(exec_env);
+        if (os_setjmp(jmpbuf_node.jmpbuf) == 0) {
```

Comment:
while OS_ENABLE_BLOCK_INSN_INTERRUPT and OS_ENABLE_HW_BOUND_CHECK set up separate jmpbufs.
signal handers seem looking at the only most recent one.

i guess it's simpler to use a single jmpbuf. how do you think?

Reply:
The function which calls setjmp in OS_ENABLE_HW_BOUND_CHECK is really complex and the whole function is controlled by OS_ENABLE_HW_BOUND_CHECK, had better not re-use the setjmp:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_runtime.c#L1239-L1335

Using another signal handler is allowed, since different signal handlers handle different signals, pthread_kill signal should be caught by this PR's new signal handler only.

Reply:
> had better not re-use the setjmp

in that case, the signal handers should not assume it's on the top of the jmpbuf stack.

Reply:
Right, another jmbuf is pushed in calling function when hw bound check is enabled, seems that we had better add a flag in the WASMJmpBuf, and in signal handler, pop jmpbuf until its flag is matched.

Reply:
They way it works now: `aot_call_function` saves the jmpbuf for instruction interruption, then the nested `invoke_native_with_hw_bound_check` save the jmpbuf for the hw bound check.
In case of `proc_exit`, the instruction interruption sig handler restores the hw bound check jmpbuf (not the corresponding instruction interruption jmpbuf) and returns to `invoke_native_with_hw_bound_check` that will in turn return to `aot_call_function`.

Not sure if we need to jump directly to `aot_call_function` as you were pointing out, I'm still debugging that since there is one case (i.e. instruction interruption and HW bound check both enabled in AOT mode) that still doesn't work correctly in my tests (the exit code is not the right one).

Reply:
Not sure, seems both ways work.
For the exit code, do you mean the wasi_proc_exit? Is `wasi_ctx->exit_code = rval;` executed after the exception was set?

Reply:
> For the exit code, do you mean the wasi_proc_exit? Is wasi_ctx->exit_code = rval; executed after the exception was set?

Exactly, I see the call to `proc_exit` but then it returns 1 at the end in that particular configuration (instruction interruption enabled and HW bound check disabled in AOT mode). I'll debug that tomorrow.

Reply:
I'm testing with the `thread_termination.c` example (updated to use `atomic.wait`) in AOT mode with HW bound check disabled.
I see another exception ("unreachable") being raised after the proc_exit and overriding the first one. Still debugging to understand what's happening there.

Reply:
Sharing the stack trace, in case I'm missing something obvious.
In AOT with HW bound check disabled, from the thread that calls `proc_exit`, after the exception set because of the `proc_exit` call,
```bash
* thread #2, name = 'iwasm', stop reason = breakpoint 8.1
  * frame #0: 0x000055555557846e iwasm`wasm_cluster_spread_exception(exec_env=0x00005555555f6710, clear=false) at thread_manager.c:1099:28
    frame #1: 0x000055555556243e iwasm`wasm_set_exception(module_inst=0x00005555555f6080, exception="wasi proc exit") at wasm_runtime_common.c:2236:9
    frame #2: 0x0000555555562531 iwasm`wasm_runtime_set_exception(module_inst_comm=0x00005555555f6080, exception="wasi proc exit") at wasm_runtime_common.c:2295:5
  
  frame #3: 0x00005555555b6365 iwasm`wasi_proc_exit(exec_env=0x00005555555f6710, rval=55) at libc_wasi_wrapper.c:989:5
```

I see another "unreachable" exception that overrides the first one and causes problems with the exit code.
(`function.func_name` = `"wasi_thread_start"`)
```bash
thread #2, name = 'iwasm', stop reason = breakpoint 9.1
    frame #0: 0x00005555555623b9 iwasm`wasm_set_exception(module_inst=0x00005555555f6080, exception="unreachable") at wasm_runtime_common.c:2224:13
    frame #1: 0x000055555556247f iwasm`wasm_set_exception_with_id(module_inst=0x00005555555f6080, id=0) at wasm_runtime_common.c:2273:9
    frame #2: 0x00005555555751ec iwasm`aot_set_exception_with_id(module_inst=0x00005555555f6080, id=0) at aot_runtime.c:1554:9
    frame #3: 0x0000555555565e9b iwasm`push_args_end at invokeNative_em64.s:61
    frame #4: 0x00007ffff7fb8a60
    frame #5: 0x0000555555564660 iwasm`wasm_runtime_invoke_native(exec_env=0x00005555555f6710, func_ptr=0x0000000040000834, func_type=0x00005555555e2fc0, signature=0x0000000000000000, attachment=0x0000000000000000, argv=0x00007ffff7fb8dc0, argc=2, argv_ret=0x00007ffff7fb8dc0) at wasm_runtime_common.c:4190:9
    frame #6: 0x0000555555574a3d iwasm`invoke_native_internal(exec_env=0x00005555555f6710, func_ptr=0x0000000040000834, func_type=0x00005555555e2fc0, signature=0x0000000000000000, attachment=0x0000000000000000, argv=0x00007ffff7fb8dc0, argc=2, argv_ret=0x00007ffff7fb8dc0) at aot_runtime.c:1344:15
    frame #7: 0x0000555555575096 iwasm`aot_call_function(exec_env=0x00005555555f6710, function=0x00005555555f66e0, argc=2, argv=0x00007ffff7fb8dc0) at aot_runtime.c:1492:15
    frame #8: 0x00005555555621ff iwasm`wasm_runtime_call_wasm(exec_env=0x00005555555f6710, function=0x00005555555f66e0, argc=2, argv=0x00007ffff7fb8dc0) at wasm_runtime_common.c:1828:15
    frame #9: 0x00005555555b19c9 iwasm`thread_start(arg=0x00005555555f6710) at lib_wasi_threads_wrapper.c:61:10
    frame #10: 0x000055555557783d iwasm`thread_manager_start_routine(arg=0x00005555555f6710) at thread_manager.c:521:11
    frame #11: 0x0000555555580c0c iwasm`os_thread_wrapper(arg=0x00005555556075e0) at posix_thread.c:43:5
    frame #12: 0x00007ffff7d25b43 libc.so.6`start_thread(arg=<unavailable>) at pthread_create.c:442:8
    
frame #13: 0x00007ffff7db7a00 libc.so.6`__clone3 at clone3.S:81

Reply:
@eloparco Could you upload the wasm file compiled so that we can also debug it? And in interrupt_block_insn_sig_handler, could you try popping the jmpbuf of hw bound check firstly, and then longjmp to the jmpbuf set by invoke_native_internal in aot_runtime.c?

Reply:
Thanks, I'll try that.

The easiest way to reproduce is to change [no_pthread.c](https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi_threads/samples/wasi-threads/wasm-apps/no_pthread.c) to
```c
/*
 * Copyright (C) 2022 Amazon.com Inc. or its affiliates. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 */
#ifndef __wasi__
#error This example only compiles to WASM/WASI target
#endif

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#include "wasi_thread_start.h"

#define STOP_FROM_MAIN 0

static const int64_t SECOND = 1000 * 1000 * 1000;

typedef struct {
    start_args_t base;
    int th_ready;
    int value;
    int thread_id;
} shared_t;

void
__wasi_thread_start_C(int thread_id, int *start_arg)
{
    shared_t *data = (shared_t *)start_arg;

    printf("New thread ID: %d, starting parameter: %d\n", thread_id,
           data->value);

#if STOP_FROM_MAIN == 1
    __builtin_wasm_memory_atomic_wait32(0, 0, -1);
#else
    sleep(2);
    __wasi_proc_exit(55);
#endif
}

int
main(int argc, char **argv)
{
    shared_t data = { { NULL }, 0, 52, -1 };
    int thread_id;
    int ret = EXIT_SUCCESS;

    if (!start_args_init(&data.base)) {
        printf("Stack allocation for thread failed\n");
        return EXIT_FAILURE;
    }

    thread_id = __wasi_thread_spawn(&data);
    if (thread_id < 0) {
        printf("Failed to create thread: %d\n", thread_id);
        ret = EXIT_FAILURE;
        goto final;
    }

#if STOP_FROM_MAIN == 1
    sleep(2);
    __wasi_proc_exit(56);
#else
    __builtin_wasm_memory_atomic_wait32(0, 0, -1);
#endif

final:
    start_args_deinit(&data.base);

    return ret;
}
```
and add `set(WAMR_DISABLE_HW_BOUND_CHECK 1)` in the [CMakeLists.txt](https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi_threads/samples/wasi-threads/CMakeLists.txt).

Reply:
> And in interrupt_block_insn_sig_handler, could you try popping the jmpbuf of hw bound check firstly, and then longjmp to the jmpbuf set by invoke_native_internal in aot_runtime.c?

Actually I have the reported problem when `WAMR_DISABLE_HW_BOUND_CHECK=1`, so that's not the cause.

Reply:
Just realized that there are problems in general with AOT. When I enabled it in #1905 I only tried with the `no_pthread.c` example. Running it again (from the HEAD of `dev/wasi_threads`) I see that the `proc_exit` doesn't behave correctly, not stopping even with non-blocking instructions.
So the problem is not specific to this PR.

I'll address the other review comments and close the PR once it's ready, AOT will be fixed in a separate PR.

Reply:
Good, how about submitting a PR to main branch firstly since the same issue also exists in main branch? I will merge main into dev/wasi_threads after several PRs are merged into main: #1924, #1929, #1902.